File: api/_lib/mcp/tools/library.ts
import type { McpToolDef } from '../protocol.js';
import { optionalBoolean, optionalInt } from '../args.js';
import { aliasIndexOf, fetchDefinitionRows, fetchLastPerformedRows } from '../data.js';
import { fetchAllPages } from '../../pagination.js';
import type { WorkoutEventRow } from '../../../../src/lib/db/types.js';
import { rowToEvent } from '../../../../src/lib/schedule/mapping.js';
import { countDefinitionReferences } from '../../../../src/lib/schedule/definitions.js';
import { lastPerformedByCanonical } from '../../../../src/lib/library/stats.js';

// Exercise-library search: lets the model resolve "incline press" to the
// canonical entry (and its aliases) before querying history.
//
// W10 widened the entries with `id` (the phone's editor PATCHes
// `/api/exercise-definitions?id=`) and an opt-in `references` count — the
// web library's "in N workouts", which it derives from every loaded event
// (countDefinitionReferences) and the phone, holding only a window of the
// schedule, cannot. The limit ceiling rose with it: the phone decorates its
// whole library in one call.

export const searchExercisesTool: McpToolDef = {
  name: 'search_exercises',
  description:
    'Search the exercise library by name or alias (case-insensitive substring). Empty query lists all. ' +
    'Returns canonical names to use with get_exercise_history, plus default prescriptions and last-performed dates. ' +
    'Set include_references true for the number of planned workouts each exercise appears in.',
  inputSchema: {
    type: 'object',
    properties: {
      query: { type: 'string', description: 'Substring to match against names and aliases. Optional.' },
      include_archived: { type: 'boolean', description: 'Include archived definitions (default false).' },
      include_references: { type: 'boolean', description: 'Count the planned workouts each exercise appears in (default false).' },
      limit: { type: 'integer', description: 'Max results (default 25, max 500).' },
    },
  },
  async run(supabase, userId, args) {
    const query = typeof args.query === 'string' ? args.query.trim().toLowerCase() : '';
    const includeArchived = optionalBoolean(args, 'include_archived', false);
    const includeReferences = optionalBoolean(args, 'include_references', false);
    const limit = optionalInt(args, 'limit', 25, 1, 500);

    const defs = await fetchDefinitionRows(supabase, userId);
    const aliasIndex = aliasIndexOf(defs);

    // Last-performed dates, aggregated server-side (one row per name per log
    // table) rather than paged through every logged (name, date) pair.
    const lastPerformed = lastPerformedByCanonical(
      await fetchLastPerformedRows(supabase, userId),
      aliasIndex.toCanonical,
    );

    // A series counts once (occurrences collapse to their base), so the base
    // rows are enough — no expansion needed.
    let references: Map<string, number> | null = null;
    if (includeReferences) {
      const eventRows = await fetchAllPages<WorkoutEventRow>('workout_events', (from, to) =>
        supabase.from('workout_events').select('*').eq('user_id', userId).order('date', { ascending: true }).range(from, to),
      );
      const events = eventRows.map(rowToEvent);
      references = new Map(defs.map(d => [d.id, countDefinitionReferences(d.id, events)]));
    }

    const matches = defs
      .filter(d => includeArchived || !d.archived_at)
      .filter(d => {
        if (!query) return true;
        if (d.canonical_name.toLowerCase().includes(query)) return true;
        return (d.aliases ?? []).some(a => a.toLowerCase().includes(query));
      })
      .slice(0, limit)
      .map(d => ({
        id: d.id,
        canonical_name: d.canonical_name,
        category: d.category,
        aliases: d.aliases ?? [],
        muscle_groups: d.muscle_groups ?? [],
        equipment: d.equipment ?? [],
        is_unilateral: d.is_unilateral,
        default_prescription: {
          sets: d.default_sets,
          reps: d.default_reps,
          duration: d.default_duration,
          weight: d.default_weight,
          rest: d.default_rest,
        },
        technique_notes: d.technique_notes ?? null,
        last_performed: lastPerformed.get(d.canonical_name) ?? null,
        archived: Boolean(d.archived_at),
        ...(references ? { references: references.get(d.id) ?? 0 } : {}),
      }));

    return { query: query || null, total_matches: matches.length, exercises: matches };
  },
};

File: ios/Packages/ApexCore/Sources/ApexCore/Models/Library.swift
import Foundation

/// `POST /api/query { tool: "search_exercises" }` — `api/_lib/mcp/tools/library.ts`.
/// Since W10 each entry carries the definition `id` and, on request, the
/// number of planned workouts it appears in (`references`) — the web
/// library's "in N workouts", counted server-side because the phone holds
/// only a window of the schedule. The definitions themselves (with defaults
/// and archive state) still come cached from `/api/schedule?include=`; this
/// call decorates that list with the stats the rows show.
public struct SearchExercisesResult: Codable, Sendable, Equatable {
    public let query: String?
    public let totalMatches: Int
    public let exercises: [Entry]

    public struct Prescription: Codable, Sendable, Equatable {
        public let sets: Int?
        public let reps: String?
        public let duration: String?
        public let weight: String?
        public let rest: String?
    }

    public struct Entry: Codable, Sendable, Equatable {
        /// Absent from a response older than W10.
        public let id: String?
        public let canonicalName: String
        public let category: String?
        public let aliases: [String]?
        public let muscleGroups: [String]?
        public let equipment: [String]?
        public let isUnilateral: Bool?
        public let defaultPrescription: Prescription?
        public let techniqueNotes: String?
        public let lastPerformed: String?
        public let archived: Bool?
        /// Only when the call asked for `include_references`.
        public let references: Int?

        enum CodingKeys: String, CodingKey {
            case id, category, aliases, equipment, archived, references
            case canonicalName = "canonical_name"
            case muscleGroups = "muscle_groups"
            case isUnilateral = "is_unilateral"
            case defaultPrescription = "default_prescription"
            case techniqueNotes = "technique_notes"
            case lastPerformed = "last_performed"
        }
    }

    enum CodingKeys: String, CodingKey {
        case query
        case totalMatches = "total_matches"
        case exercises
    }

    /// The library's decoration read: everything, archived included, with
    /// the reference counts. The ceiling is the server's.
    public static func args(query: String = "", includeArchived: Bool = true, includeReferences: Bool = true, limit: Int = 500) -> [String: JSONValue] {
        [
            "query": .string(query),
            "include_archived": .bool(includeArchived),
            "include_references": .bool(includeReferences),
            "limit": .number(Double(limit)),
        ]
    }
}

/// `POST /api/query { tool: "get_exercise_history" }` — `api/_lib/mcp/tools/tracking.ts`.
/// Alias-aware: a former name resolves to the current one and `resolvedFrom`
/// says so. An exercise with no logged history is a 400, not an empty result.
public struct ExerciseHistoryResult: Codable, Sendable, Equatable {
    public struct Best: Codable, Sendable, Equatable {
        public let display: String
        public let date: String
    }

    public struct TrendPoint: Codable, Sendable, Equatable {
        public let date: String
        public let value: Double
    }

    public struct Session: Codable, Sendable, Equatable {
        public let date: String
        /// Logged sets in order, e.g. "185 × 5" or "0:45".
        public let sets: [String]
    }

    public let canonicalName: String
    public let resolvedFrom: String?
    /// `oneRM`, `duration`, `reps`, `distance`, or nil when nothing parsed.
    public let statKind: String?
    /// The axis and PR label: "est. 1RM", "hold", "reps", or the cardio unit.
    public let statUnit: String
    public let allTimeBest: Best?
    public let totalSessions: Int
    public let trend: [TrendPoint]
    public let recentSessions: [Session]

    enum CodingKeys: String, CodingKey {
        case trend
        case canonicalName = "canonical_name"
        case resolvedFrom = "resolved_from"
        case statKind = "stat_kind"
        case statUnit = "stat_unit"
        case allTimeBest = "all_time_best"
        case totalSessions = "total_sessions"
        case recentSessions = "recent_sessions"
    }

    public static func args(exerciseName: String, limit: Int = 10) -> [String: JSONValue] {
        ["exercise_name": .string(exerciseName), "limit": .number(Double(limit))]
    }
}
